File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
#include <iostream>

class Solution {
public:
    int myAtoi(const std::string& str) {
        int index = 0;
        return recursiveAtoi(str, index);
    }

private:
    int recursiveAtoi(const std::string& str, int& index) {
        // Skip leading whitespaces
        while (index < str.length() && str[index] == ' ') {
            index++;
        }

        // Check for sign
        int sign = 1;
        if (index < str.length() && (str[index] == '+' || str[index] == '-')) {
            sign = (str[index++] == '-') ? -1 : 1;
        }

        // Calculate the result
        long long result = 0;
        while (index < str.length() && isdigit(str[index])) {
            result = result * 10 + (str[index++] - '0');

            // Check for overflow
            if (result * sign > INT_MAX) {
                return INT_MAX;
            } else if (result * sign < INT_MIN) {
                return INT_MIN;
            }
        }

        return static_cast<int>(result * sign);
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        long num = 0;
        char[] a = s.toCharArray();
        int i = 0;
        int sign = 1; // Initialize to 1 for positive numbers
        char current;

        // Skip leading whitespace
        while (i < a.length && a[i] == ' ') {
            i++;
        }

        // Check for sign if any
        if (i < a.length && a[i] == '+') {
            sign = 1;
            i++;
        } else if (i < a.length && a[i] == '-') {
            sign = -1;
            i++;
        }

        // Convert characters to number
        while (i < a.length && a[i] >= '0' && a[i] <= '9') {
            current = a[i];
            num = num * 10 + (current - '0');

            // Check for overflow or underflow
            if (num * sign <= Integer.MIN_VALUE) {
                return Integer.MIN_VALUE;
            } else if (num * sign >= Integer.MAX_VALUE) {
                return Integer.MAX_VALUE;
            }
            i++;
        }

        return (int) (num * sign);
    }
}